archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,718 @@
+  begin
+    intro k,
+    apply le_of_lt,
+    rw one_div_pos,
+    rw sq_pos_iff,
+    exact nonzero_of_invertible ↑(succ k),
+  end,
+  have g:= partial_sum_le_tsum summable_inverse_squares h,
+  cases g with d g,
+  use d,
+  intro n,
+  exact g (range n),
+end
+
+
+/-!
+ ### Part 1
+ https://proofwiki.org/wiki/Stirling%27s_Formula#Part_1
+-/
+
+/--
+Define `stirling_seq n` as $\frac{n!}{\sqrt{2n}/(\frac{n}{e})^n$.
+Stirling's formula states that this sequence has limit $\sqrt(π)$.
+-/
+noncomputable def stirling_seq (n : ℕ) : ℝ :=
+(n.factorial : ℝ) / ((sqrt(2 * n) * ((n / (exp 1))) ^ n))
+
+/-- The function `log(1 + x) - log(1 - x)` has a power series expansion with k-th term
+`2 * x^(2 * k + 1) / (2 * k + 1)`, valid for `|x| < 1`. -/
+lemma log_sum_plus_minus (x : ℝ) (hx : |x| < 1) : has_sum (λ k : ℕ,
+  (2 : ℝ) * (1 / (2 * (k : ℝ) + 1)) * (x ^ (2 * k + 1))) (log (1 + x) - log(1 - x)) :=
+begin
+ have h₁, from has_sum_pow_div_log_of_abs_lt_1 hx,
+  have h₂, from has_sum_pow_div_log_of_abs_lt_1 (eq.trans_lt (abs_neg x) hx),
+  replace h₂ := (has_sum_mul_left_iff  (λ h : ( -1 = (0:ℝ)), one_ne_zero $ neg_eq_zero.mp h)).mp h₂,
+  rw [neg_one_mul, neg_neg, sub_neg_eq_add 1 x] at h₂,
+  have h₃, from has_sum.add h₂ h₁,
+  rw [tactic.ring.add_neg_eq_sub] at h₃,
+  let term := (λ n :ℕ, ((-1) * ((-x) ^ (n + 1) / ((n : ℝ) + 1)) + (x ^ (n + 1) / ((n : ℝ) + 1)))),
+  let two_mul := (λ (n : ℕ), (2 * n)),
+  rw ←function.injective.has_sum_iff (mul_right_injective two_pos) _ at h₃,
+  { suffices h_term_eq_goal :
+    (term ∘ two_mul) = (λ k : ℕ, 2 * (1 / (2 * (k : ℝ) + 1)) * x ^ (2 * k  + 1)),
+    by {rw h_term_eq_goal at h₃, exact h₃},
+    apply funext,
+    intro n,
+    rw [function.comp_app],
+    dsimp only [two_mul, term],
+    rw odd.neg_pow (⟨n, rfl⟩ : odd (2 * n + 1)) x,
+    rw [neg_one_mul, neg_div, neg_neg, cast_mul, cast_two],
+    ring },
+  { intros m hm,
+    rw [range_two_mul, set.mem_set_of_eq] at hm,
+    rw [even.neg_pow (even_succ.mpr hm), succ_eq_add_one, neg_one_mul, neg_add_self] },
+end
+
+/--
+For any natural number `n ≠ 0`, we have the identity
+`log((n + 1) / n) = log(1 + 1 / (2*n + 1)) - log(1 - 1/(2*n + 1))`.
+-/
+lemma log_succ_div_eq_log_sub (n : ℕ) (hn : n ≠ 0) :
+  log (n.succ / n) = log (1 + 1 / (2 * n + 1)) - log (1 - 1 / (2 * n + 1)) :=
+begin
+  have : (2 : ℝ) * n + 1 ≠ 0, by { norm_cast, exact (2 * n).succ_ne_zero, },
+  rw ← log_div,
+  suffices h, from congr_arg log h,
+    all_goals {field_simp [cast_ne_zero.mpr]}, /- all_goals prevents using brackets here -/
+  { ring },
+  { norm_cast, exact succ_ne_zero (2 * n + 1) },
+end
+
+/--
+For any natural number `n`, the expression `log((n + 1) / n)` has the series expansion
+$\sum_{k=0}^{\infty}\frac{2}{2k+1}(\frac{1}{2n+1})^{2k+1}$.
+-/
+lemma power_series_log_succ_div (n : ℕ) (hn : n ≠ 0) : has_sum (λ (k : ℕ),
+  (2 : ℝ) * (1 / (2 * (k : ℝ) + 1)) * ((1 / (2 * (n : ℝ) + 1)) ^ (2 * k + 1)))
+  (log ((n.succ : ℝ) / (n : ℝ))) :=
```

Comment:
This lemma (and the two previous) hold for positive real `n`, not just natural numbers `n`.  Stating it in this way will postpone the casting difficulties until they are really relevant, and it will also make this lemma a better candidate for inclusion in the main library, near `has_sum_pow_div_log_of_abs_lt_1`.

(I'd probably rename from `n` to something more suggestive of a real number.)

Reply:
Done (it only touched this lemma and the previous one, in the one before that we already had everything real)

I you think it is a good idea, I can factor out `power_series_log_succ_div` and it's two prerequisites into `analysis.special_functions.log.deriv`.